.buildkite/filter-pipeline.py

```diff
@@ -0,0 +1,87 @@
+#! /usr/bin/env python3
+
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+import argparse, re
+
+from yaml import FullLoader, load, dump
+
+
+def check_step(step, diff):
+    if "only_changes" not in step:
+        return True
+    else:
+        changes_regexes = step["only_changes"]
+        for regex in changes_regexes:
+            for diff_file in diff:
+                r = re.compile(regex)
+                match = r.match(diff_file)
+                if match is not None:
+                    return True
+        return False
+
+
+def remove_only_changes(steps):
+    for step in steps:
+        if type(step) != str:
+            step.pop("only_changes", None)
+    return steps
+
+
+def dfs(name, adjacency_dict, visited):
+    if name not in visited:
+        visited.add(name)
+        for adjacent in adjacency_dict.get(name, []):
+            dfs(adjacent, adjacency_dict, visited)
+
+
+def build_tree(steps):
+    adjacency_dict = {}
+    adjacency_dict_rev = {}
+    for step in steps:
+        label = step["label"]
+        depends_on = step.get("depends_on", None)
+        if depends_on is not None:
+            if type(depends_on) == str:
+                adjacency_dict.setdefault(depends_on, []).append(label)
+                adjacency_dict_rev.setdefault(label, []).append(depends_on)
+            else:
+                for dependency in depends_on:
+                    if type(dependency) == str:
+                        dependency_name = dependency
+                    else:
+                        dependency_name = dependency["step"]
+                    adjacency_dict.setdefault(dependency_name, []).append(label)
+                    adjacency_dict_rev.setdefault(label, []).append(dependency_name)
+    return adjacency_dict, adjacency_dict_rev
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--pipeline", required=True)
+parser.add_argument("--git-diff", required=True)
+parser.add_argument("--output", required=True)
+args = parser.parse_args()
+
+splitted_diff = args.git_diff.strip().split("\n")
+pipeline = load(open(args.pipeline, "r"), Loader=FullLoader)
+
+visited = set()
+visited_rev = set()
+steps = pipeline["steps"]
+adjacency_dict, adjacency_dict_rev = build_tree(steps)
+triggered_steps = filter(lambda x: check_step(x, splitted_diff), steps)
+for step in triggered_steps:
+    label = step["label"]
+    dfs(label, adjacency_dict, visited)
+    dfs(label, adjacency_dict_rev, visited_rev)
+
+for step in steps:
+    label = step["label"]
+    if (label not in visited) and (label not in visited_rev):
```

Comment:
I guess it would be better to check `not in` the union of the sets.

Reply:
:open_mouth: well that's surprising, but on a quick research it seems that's indeed the case

For reference, apparently the time-complexity of `in` is O(1) on average and O(N) at worst whereas set union is O(n + m) on average.